back-end/routes/users.js

```diff
@@ -1,55 +1,169 @@
+const { pool } = require('../config');
+const {check, validationResult} = require('express-validator');
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const expjwt = require('express-jwt');
 
-const makeDbQueryAndReturnResults = (queryString, res) => {
-  pool.query(queryString, (error, results) => {
-    if (!results || !results.rows || results.rows[0] === undefined || error) {
-      return res.status(500).send('An error occurred');
-    }
-    return res.status(200).send(results.rows);
-  })
-}
+const PASSWORD_MIN_LENGTH = 8;
 
-const makeDbQueryAndReturnStatus = (queryString, res) => {
+const makeDbQueryAndReturnResults = (queryString, res) => {
   pool.query(queryString, (error, results) => {
     if (error) {
+      return res.status(500).send(error);
+    }
+    else if (!results || !results.rows || results.rows[0] === undefined) {
       return res.status(500).send('An error occurred');
     }
-    return res.status(200);
+    return res.status(200).send(results.rows);
   })
-}
+};
 
-/* GET all users */
-router.get('/', function(req, res, next) {
-  const getAllUsersQuery = 'SELECT * FROM users';
-  makeDbQueryAndReturnResults(getAllUsersQuery, res);
-});
+// async function getRowFromDb(queryString) {
+//   let results;
+//   try {
+//     results = await pool.query(queryString)
+//   } catch (e) {
+//     throw e
+//   }
+//   return results.rows[0];
+// };
 
+// TODO: Need to validate the user has access to this query
 /* GET user by ID */
 router.get('/:id', function(req, res, next) {
   const getUserQuery = 'SELECT * FROM users WHERE id=' + req.params.id;
   makeDbQueryAndReturnResults(getUserQuery, res);
 });
 
-/* POST to create new users */
-router.post('/', function(req, res, next) {
-  const {
-    username, password, github_username
-  } = req.body;
-
-  const createUserQuery = `INSERT INTO users(username, password, github_username, created_at) \
-VALUES('${username}', '${password}', '${github_username}', CURRENT_TIMESTAMP) returning *`
-  console.log(createUserQuery)
-  makeDbQueryAndReturnResults(createUserQuery, res);
+// UPDATED to be async for bcrypt library.
+router.post('/login', async function (req, res, next) {
+
+  // ie with a hashed password stored and checking against that.
+  try {
+    const query = "SELECT password, id FROM users WHERE email = ?";
+
+    const { email, password } = req.body;
+    let userID;
+    // initialize as invalid password value.
+    let passwordCheck = "bad";
+
+    // Not sure if this is the best way to do this query, maybe
+    // it should be a separate function.
+    pool.query(query, email, function(error, results, fields){
+      if(error){
+        res.write(JSON.stringify(error));
+        res.end();
+      }
+      // value changes if query worked.
+      passwordCheck = results[0];
+      userID = results[1];
+    });
+
+    if(await bcrypt.compare(password, passwordCheck)){
+      const token = jwt.sign({ userID }, process.env.ACCESS_TOKEN_SECRET);
```

Comment:
@AndreEPaul did you already add this ACCESS_TOKEN_SECRET to the heroku server as an environment variable?

Reply:
I have not, wanted to ask if we already had a dotenv on the heroku server

Reply:
I haven't set one up yet.

Reply:
There are some environment variables that I know are working for the postgres db on heroku, but those are not set by us in dotenv file.